### ISSUE TITLE
Headless Analyzer has partially completely different results

### ISSUE THREAD

**Issue by MrMatch246** [NationalSecurityAgency/ghidra#5531]
**Describe the bug**
Im currently trying to use the headless analyzer on binaries from binutils (they have been un-stripped with eu-unstrip , and in this case also stripped again)
I am talking about the size binary of binutils-multiarch_2.31.1-16_amd64

i looked at it in ghidra gui, everything is fine, then with the analyzeHeadless mode and there it basically lacks all functions.
im trying to export them with a post script. which i can [link](https://github.com/MrMatch246/rAIversing/blob/main/modules/ghidra/ghidra_scripts/ExtractCcode.py) here.

i noticed after some debugging that the headless analyzer assumed  _fastcall whereas the gui used _stdcall
Changing that with a prescript didnt fix it.

Expected (and in GUI):

```
void processEntry entry(undefined8 param_1,undefined8 param_2)

{
  undefined auStack_8 [8];
  
  __libc_start_main(FUN_001024b0,param_2,&stack0x00000008,FUN_00104ff0,FUN_00105050,param_1,
                    auStack_8);
  do {
                    /* WARNING: Do nothing block with infinite loop */
  } while( true );
}
```

Actual (as exported via headless):

void entry(undefined8 param_1,undefined8 param_2)

```
{
  undefined auStackX_8 [32];
  
  (*(code *)PTR___libc_start_main_00108fd8)(&LAB_00104ff0,auStackX_8,&DAT_00105050,param_2);
  do {
                    /* WARNING: Do nothing block with infinite loop */
  } while( true );
}
```

```
x86 Constant Reference Analyzer=true
Data Reference.Relocation Table Guide=true
ASCII Strings=true
Shared Return Calls.Allow Conditional Jumps=false
Create Address Tables.Allow Offcut References=false
Reference.Unicode String References=true
Decompiler Parameter ID.Analysis Decompiler Timeout (sec)=60
Function ID.Create Analysis Bookmarks=true
Create Address Tables=true
Disassemble Entry Points.Respect Execute Flag=true
Aggressive Instruction Finder=false
Reference.Address Table Minimum Size=2
DWARF=true
ASCII Strings.Minimum String Length=LEN_5
Reference.Minimum String Length=5
Function ID=true
Non-Returning Functions - Known.Create Analysis Bookmarks=true
Reference.Ascii String References=true
Stack.useNewFunctionStackAnalysis=true
Reference.Relocation Table Guide=true
Demangler GNU.Apply Function Signatures=true
Variadic Function Signature Override.Create Analysis Bookmarks=false
Function Start Search=true
x86 Constant Reference Analyzer.Min absolute reference=4
Data Reference.Address Table Alignment=1
Decompiler Parameter ID=false
Non-Returning Functions - Discovered.Repair Flow Damage=true
Apply Data Archives.Create Analysis Bookmarks=true
Create Address Tables.Create Analysis Bookmarks=true
Reference.References to Pointers=true
Apply Data Archives.Archive Chooser=[Auto-Detect]
ELF Scalar Operand References=true
Non-Returning Functions - Discovered=true
Apply Data Archives.User Project Archive Path=
Create Address Tables.Relocation Table Guide=true
Reference.Address Table Alignment=1
Data Reference.Align End of Strings=false
Function ID.Always Apply FID Labels=false
DWARF.Output Source Info=false
Apply Data Archives=true
ASCII Strings.Create Strings Containing Existing Strings=true
DWARF.Preload All DIEs=false
Data Reference.Create Address Tables=false
ASCII Strings.Force Model Reload=false
Function ID.Multiple Match Threshold=30.0
x86 Constant Reference Analyzer.Function parameter/return Pointer analysis=true
Data Reference.Ascii String References=true
Data Reference.Respect Execute Flag=true
Data Reference.Subroutine References=true
Function ID.Instruction Count Threshold=14.6
Shared Return Calls=true
Data Reference=true
Condense Filler Bytes=false
Reference=true
Subroutine References=true
x86 Constant Reference Analyzer.Create Data from pointer=false
ASCII Strings.Create Strings Containing References=true
Disassemble Entry Points=true
Aggressive Instruction Finder.Create Analysis Bookmarks=true
ELF Scalar Operand References.Relocation Table Guide=true
Demangler GNU=true
Shared Return Calls.Assume Contiguous Functions Only=true
Data Reference.Switch Table References=false
Condense Filler Bytes.Filler Value=Auto
DWARF.Add Lexical Block Comments=false
Stack.Create Local Variables=true
Call Convention ID.Analysis Decompiler Timeout (sec)=60
Data Reference.Address Table Minimum Size=2
Create Address Tables.Table Alignment=4
DWARF.Output DWARF DIE Info=false
Demangler GNU.Use Deprecated Demangler=false
Data Reference.Unicode String References=true
Embedded Media.Create Analysis Bookmarks=true
Decompiler Parameter ID.Prototype Evaluation=__fastcall
ASCII Strings.String Start Alignment=ALIGN_1
Decompiler Parameter ID.Analysis Clear Level=ANALYSIS
Decompiler Switch Analysis.Analysis Decompiler Timeout (sec)=60
ASCII Strings.Require Null Termination for String=true
Decompiler Parameter ID.Commit Data Types=true
x86 Constant Reference Analyzer.Trust values read from writable memory=true
Create Address Tables.Minimum Table Size=2
Data Reference.References to Pointers=true
Demangler GNU.Apply Function Calling Conventions=true
Non-Returning Functions - Discovered.Create Analysis Bookmarks=true
Create Address Tables.Pointer Alignment=1
External Entry References=true
DWARF.Add Inlined Functions Comments=false
Reference.Respect Execute Flag=true
x86 Constant Reference Analyzer.Require pointer param data type=false
ASCII Strings.String end alignment=4
x86 Constant Reference Analyzer.Stored Value Pointer analysis=true
DWARF.Debug Item Limit=2000000
Non-Returning Functions - Known=true
Stack=true
Reference.Align End of Strings=false
Non-Returning Functions - Discovered.Function Non-return Threshold=3
Reference.Subroutine References=true
Subroutine References.Create Thunks Early=true
Call Convention ID=true
Demangler GNU.Demangler Format=AUTO
ASCII Strings.Search Only in Accessible Memory Blocks=true
Reference.Create Address Tables=false
ASCII Strings.Model File=StringModel.sng
Demangler GNU.Demangle Only Known Mangled Symbols=false
x86 Constant Reference Analyzer.Speculative reference max=512
Embedded Media=true
Reference.Switch Table References=false
Data Reference.Minimum String Length=5
x86 Constant Reference Analyzer.Max Threads=2
Create Address Tables.Minimum Pointer Address=4132
DWARF.Import Functions=true
Decompiler Switch Analysis=true
Decompiler Parameter ID.Commit Void Return Values=false
DWARF.Maximum Name Length=2000
Function Start Search.Search Data Blocks=false
x86 Constant Reference Analyzer.Speculative reference min=1024
DWARF.Try To Pack Structs=true
Call-Fixup Installer=true
DWARF.Import Data Types=true
Create Address Tables.Auto Label Table=false
Create Address Tables.Maxmimum Pointer Distance=16777215
DWARF.Create Function Signatures=true
Stack.Create Param Variables=true
Function Start Search.Bookmark Functions=false
Condense Filler Bytes.Minimum number of sequential bytes=1
```

**To Reproduce**
/home/.../ghidra_10.3_PUBLIC/support/analyzeHeadless /home/.../testing -import /home/.../testing/stripped/x86_64-linux-gnu-size -preScript TestPreScript.py -postScript ExtractCcode.py 
"<arg for extraction>" -scriptPath /home/.../ghidra_scripts -scriptlog /home/.../scriptlog -log /home/.../projects/log -processor x86:LE:64:default -max-cpu 6

**Environment (please complete the following information):**
 - OS: uname -r  : 4.19.0-21-amd64
 - Java Version: OpenJDK Runtime Environment (build 20+36-2344)
OpenJDK 64-Bit Server VM (build 20+36-2344, mixed mode)
 - Ghidra Version:10.3 (same in 10.3.1)
 - Ghidra Origin: github release

FILE: 
[x86_64-linux-gnu-size.zip](https://github.com/NationalSecurityAgency/ghidra/files/11967679/x86_64-linux-gnu-size.zip)

**Comment by MrMatch246** [comment]
I encountered this with multiple different binaries now the output from analyzeHeadless is at times really far from the gui.
Id be open for every bit of help here.

**Comment by MrMatch246** [comment]
TLDR
i only supplied the language ID , now the headlessAnalyzer assumed compilerId windows while the gui assumed gcc.
if you are doing the same either supply both or none.

**Comment by MrMatch246** [comment]
Well either you supplie languageID and compilerID to the headlessAnalyzer or neither and let it find it out itself.
Basically you only supply them if it cant find them out itself but when you do you supply both.
I hope that helps. 

Am 25. November 2024 18:00:31 MEZ schrieb Ada Lindberg ***@***.***>:
>> TLDR i only supplied the language ID , now the headlessAnalyzer assumed compilerId windows while the gui assumed gcc. if you are doing the same either supply both or none.
>
>I'm having a similar issue with getting inconsistent output between running post-processing code in the context of a GhidraScript or in the same context as the headless analyzer.
>
>Could you elaborate what do you mean by "if you are doing the same either supply both or none."?
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/NationalSecurityAgency/ghidra/issues/5531#issuecomment-2498559811
>You are receiving this because you modified the open/close state.
>
>Message ID: ***@***.***>
-- 
Diese Nachricht wurde von meinem Android-Gerät mit K-9 Mail gesendet.
